lang/syn/src/parser/accounts/mod.rs

```diff
@@ -39,55 +39,89 @@ pub fn parse(strct: &syn::ItemStruct) -> ParseResult<AccountsStruct> {
+                    return Err(ParseError::new(
+                        init_fields[0].ident.span(),
+                        "the init constraint requires \
                             the token_program field to exist in the account \
                             validation struct. Use the program type to add \
                             the token_program field to your validation struct.",
-                        ));
-                    }
+                    ));
                 }
             }
-            // a_token needs associated token program.
-            if let InitKind::AssociatedToken { .. } = kind {
-                if fields
-                    .iter()
-                    .all(|f| f.ident() != "associated_token_program")
-                {
-                    return Err(ParseError::new(
-                        init_field.ident().span(),
-                        "the init constraint requires \
+        }
+        // a_token needs associated token program.
+        if let InitKind::AssociatedToken { .. } = kind {
+            if fields
+                .iter()
+                .all(|f| f.ident() != "associated_token_program")
+            {
+                return Err(ParseError::new(
+                    init_fields[0].ident.span(),
+                    "the init constraint requires \
                     the associated_token_program field to exist in the account \
                     validation struct. Use the program type to add \
                     the associated_token_program field to your validation struct.",
+                ));
+            }
+        }
+
+        for field in init_fields {
+            // Get payer for init-ed account
+            let associated_payer_name = match field.constraints.init.clone().unwrap().payer.unwrap()
+            {
+                Expr::Path(expr) if expr.path.segments.len() == 1 => {
+                    expr.path.segments[0].ident.clone()
+                }
+                _ => {
+                    // composite payer, check not supported
+                    continue;
+                }
+            };
+
+            // Check payer is mutable
+            let associated_payer_field = fields.iter().find_map(|f| match f {
+                AccountField::Field(field) if *f.ident() == associated_payer_name => Some(field),
+                _ => None,
+            });
+            match associated_payer_field {
+                Some(associated_payer_field) => {
+                    if !associated_payer_field.constraints.is_mutable() {
+                        return Err(ParseError::new(
+                            field.ident.span(),
+                            "the payer specified for an init constraint must be mutable.",
+                        ));
+                    }
+                }
+                _ => {
+                    return Err(ParseError::new(
+                        field.ident.span(),
+                        "the init constraint requires a payer.",
```

Comment:
The problem is not that `init` does not have a payer but that the given payer doesn't exist.